### ISSUE TITLE
using diffsort for ranking

### ISSUE THREAD

**Issue by kiranchari** [Felix-Petersen/diffsort#6]
I would like to use diffsort to rank a pytorch tensor minibatch of shape (N,D), along the second dimension. Could you please provide an example of how to obtain the ranking? I am not sure how to use permutation_matrix. 

Also, what are the best values for the hyperparameters (steepness and art_lambda)? I noticed that sorted_vectors often contains values close to but not equal to those in the input array. I understand that increasing steepness makes the approximation more discrete and closer to the original values but is there a right way to set the hyperparameter? 

Also, I noticed that the gradient values are all 1 when i tried sorting a list of random values, which I don't think is accurate.

**Comment by Felix-Petersen** [comment]
The `permutation_matrix` indicates the probability of each input to be at a certain rank. That means, `permutation_matrix` encodes / is the ranking. If you could go into more detail on what your goal is, I am happy to provide more details. In case you need a real-valued ranking rather than a probability of rankings, the matrix product between the permutation matrix and `torch.arange(n)` should work. However, for the applications I have encountered so far, using the probabilities, (i.e., `permutation_matrix`) is much better.

For the distribution, I recommend using `cauchy`, in which case `art_lambda` is irrelevant (that is only relevant for the logistic art). For `steepness`, a value of 10 often works well, but it is important to explore the steepness on a larger grid, e.g. [0.1, 1, 10, 100, 1000], and then finer grained testing. In particular, it depends a lot on the magnitude / std of the inputs and experiment setting. 

If you did `sorted_vectors.sum().backward()`, then all gradients being 1 should be expected. (Think: if the task is to increase all each sorted output by 1, then one needs to increase every input by 1.)

**Comment by kiranchari** [comment]
Thanks for your response @Felix-Petersen. I obtained the ranks using `ranks = permutation_matrices @ torch.arange(n).float()`. I used `DiffSortNet('bitonic', N, steepness=1000)` for `N=128`, which produced a good soft ranking. (And you're right that `sorted_vectors.sum().backward()` should result in gradients being 1.)
